// @flow/src/compiler/templateCompiler.js
import { evaluateJSExpression } from '../../utils/evaluateJSExpression.js'
import { replaceInterpolationMarkers } from '../../utils/formatTemplate.js'
import { bindForDirective } from './../compiler/directives/for.js'
import { bindIfDirective } from './../compiler/directives/if.js'
import { mountStyling } from '../../src/compiler/mountStyling.js'
import { getTranslatedAttributeNames, translateBindingAttribute, translateEventAttribute } from '../../utils/getTranslatedAttributeNames.js'
import { reactiveVariable } from '../../reactivity/reactiveVariable.js'
import {
    REACTIVE_VARIABLE_REF_CONFIG,
    REACTIVE_VARIABLE_COMPUTED_CONFIG,
    REACTIVE_VARIABLE_PROP_CONFIG,
    LIFECYCLE_CALLBACKS_TEMPLATE
} from '../../configuration/configuration.js'

const camelizeKebabCasedString = string => string.replace(/-./g, character => character[1].toUpperCase())

export function compileComponent(componentFunction = () => { }, componentParentNode = {}, parentContext = {}) {
    componentParentNode.$FlowComponent ??= {}
    componentParentNode.$FlowComponent.properties ??= {}
    const upperCaseComponentName = componentFunction.name.toUpperCase()

    const lifecycleCallbacks = structuredClone(LIFECYCLE_CALLBACKS_TEMPLATE)
    const properties = componentParentNode.$FlowComponent.properties

    // Defining component's destructurable properties.
    const componentContextMethods = {
        emit: (eventName, payload, eventInitDict = { bubbles: true }) => {
            eventInitDict.detail = payload

            const customEvent = new CustomEvent(eventName, eventInitDict)

            componentParentNode.dispatchEvent(customEvent)
        },
        ref: (value) => {
            return reactiveVariable(value, structuredClone(REACTIVE_VARIABLE_REF_CONFIG))
        },
        computed: (callback) => {
            return reactiveVariable(callback, structuredClone(REACTIVE_VARIABLE_COMPUTED_CONFIG))
        },
        useStore: (storeName) => {
            const store = $FlowEngine.useStore(storeName)

            return store
        },
        useRouter: () => {
            const routerStore = $FlowEngine.useStore('routerStore')

            return routerStore
        },
        watch: (source, callback, watchConfiguration) => {
            if (!source.configuration?.isReactiveVariable) return

            const variableReference = source.variableReference

            if (watchConfiguration?.immediate) {
                callback(source.value, source.value)
            }

            $FlowEngine.watchCallbacks[variableReference].push(callback)
        },
        onMounted: (callback) => {
            lifecycleCallbacks.onMounted = callback
        },
        onBeforeMount: (callback) => {
            lifecycleCallbacks.onBeforeMount = callback
        },
        props: componentParentNode.$FlowComponent.properties
    }

    // Binding props and transforming them into reactive variables.
    let componentProperties = $FlowEngine.propertiesByComponent[componentParentNode.tagName] || [] //  - in case the parent node's tagName is 'DIV'.

    componentProperties.forEach((propertyName) => {
        const attributeExpression = componentParentNode.getAttribute(`data-flow-attribute-original-content-${propertyName}`)

        const camelCasePropertyName = camelizeKebabCasedString(propertyName)

        const propertyReactiveVariable = reactiveVariable(undefined, structuredClone(REACTIVE_VARIABLE_PROP_CONFIG))

        properties[camelCasePropertyName] = propertyReactiveVariable

        $FlowEngine.queueReactiveEffect({
            effect: () => {
                propertyReactiveVariable.value = evaluateJSExpression(parentContext, attributeExpression)
            }
        })
    })

    // Calling a component function with the context methods. The component function returns a context object.
    // It is fine to think of it as Vue's onCreated lifecycle hook.
    const stringifiedComponentFunction = componentFunction.toString()

    const evaluatedComponentFunction = evaluateJSExpression(componentContextMethods, stringifiedComponentFunction)

    const componentContext = Object.assign(properties, evaluatedComponentFunction())

    // Setting properties on componentContext that can be later referenced in templates by preprending them with '$' sign.
    componentContext.$path = $FlowEngine.useStore('routerStore').path
    componentContext.$emit = componentContextMethods.emit


    // Setting componentParentNode's innerHTML to the component's template and removing the 'template' property.
    const componentTemplate = $FlowEngine.templateByComponent[upperCaseComponentName] || ''

    // Invoking onBeforeMount lifecycle hook if defined.
    lifecycleCallbacks.onBeforeMount?.()

    // ! After running this instruction component gets its HTML content set.
    componentParentNode.innerHTML = replaceInterpolationMarkers(componentTemplate)

    const componentStyle = $FlowEngine.styleByComponent[upperCaseComponentName] || ''

    // Grabbing all inner elements of the componentParentNode that has just been created by innerHTML property.
    const componentNodeInnerElements = [...componentParentNode.querySelectorAll('*:not([data-flow-compiled-component])')]

    // Formatting every inner node in order for easier manipulation.
    let mappedInnerNodes = componentNodeInnerElements.map((element) => {
        const attributes = [...element.attributes]

        return {
            element,
            attributes: getTranslatedAttributeNames(attributes)
        }
    })

    // This is so-called "loop-marking". It is used to distinguish looped elements from the rest.
    // This because loops are mounted separately from the rest of the component.
    const loopNodes = [...componentParentNode.querySelectorAll('Loop')]

    loopNodes.forEach(element => {
        const loopBlockElements = [...element.querySelectorAll('*')]

        loopBlockElements.forEach((loopBlockElement) => {
            loopBlockElement.setAttribute('data-flow-looped-element', true)
        })
    })

    // Determining component drawers.
    const drawerNodes = [...componentParentNode.querySelectorAll('Drawer')]
    const mappedInnerDrawerTags = drawerNodes.map(drawerNode => {
        const drawerName = drawerNode.getAttribute("name")

        const targetInsertTag = componentParentNode.$FlowComponent.mappedInnerInsertTags.find(insertTag => insertTag.name === drawerName)

        if (targetInsertTag) {
            drawerNode.prepend(targetInsertTag.insertNode)
        }

        return {
            element: drawerNode,
            name: drawerNode.getAttribute("name")
        }
    })

    componentParentNode.$FlowComponent.mappedInnerDrawerTags = mappedInnerDrawerTags

    const componentNamesUsed = []

    // The famous "mounting" step. This is where the magic happens.
    mappedInnerNodes.forEach((mappedInnerNode) => {
        const isComponentModule = $FlowEngine.componentModules[mappedInnerNode.element.tagName]

        // * insertTags
        if (isComponentModule) {
            componentNamesUsed.push(mappedInnerNode.element.tagName)

            mappedInnerNode.element.$FlowComponent = {}

            const insertTags = [...mappedInnerNode.element.querySelectorAll('Insert')]
            const mappedInnerInsertTags = insertTags.map(insertNode => {
                return {
                    componentNode: mappedInnerNode.element,
                    insertNode,
                    name: insertNode.getAttribute("name")
                }
            })

            mappedInnerNode.element.$FlowComponent.mappedInnerInsertTags = mappedInnerInsertTags
        }

        // * eventAttributes
        // Mounting event attributes that start with '@' character.
        mappedInnerNode.attributes.eventAttributes.forEach(({ name: eventAttribute, value }) => {
            const evaluatedCallaback = evaluateJSExpression(componentContext, value)

            mappedInnerNode.element.addEventListener(translateEventAttribute(eventAttribute), evaluatedCallaback)
        })

        // * bindingAttributes
        // Mounting binding attributes that start with ':' character.
        mappedInnerNode.attributes.bindingAttributes.forEach(({ name: bindingAttribute, value }) => {
            const translatedBindingAttribute = translateBindingAttribute(bindingAttribute)

            mappedInnerNode.element.setAttribute(`data-flow-attribute-original-content-${translatedBindingAttribute}`, value)

            // Note: for loop is mounted separately from other elements and has its own mounting (copied) step.
            if (bindingAttribute === ':for') {
                bindForDirective(mappedInnerNode, componentContext)

                return
            }

            const isNodeNotAChildOfLoopElement = !mappedInnerNode.element.getAttribute('data-flow-looped-element')

            // Mounting :if directive.
            if (isNodeNotAChildOfLoopElement) {
                if (bindingAttribute === ':if') {
                    bindIfDirective(mappedInnerNode, componentContext)

                    return
                }

                // Mounting input elements' value and checked attributes.
                const inputTypes = ['checkbox', 'radio', 'text', 'number', 'password', 'email', 'url', 'tel', 'date', 'time', 'color', 'range'];

                inputTypes.forEach(inputType => {
                    const isNodeAInput =
                        mappedInnerNode.element.tagName === 'INPUT' &&
                        mappedInnerNode.element.getAttribute('type') === inputType;

                    if (isNodeAInput) {
                        $FlowEngine.queueReactiveEffect({
                            effect: () => {
                                const condition = evaluateJSExpression(componentContext, value);

                                if (inputType === 'checkbox' || inputType === 'radio') {
                                    if (condition) {
                                        mappedInnerNode.element.setAttribute('checked', condition);
                                    } else {
                                        mappedInnerNode.element.removeAttribute('checked');
                                    }
                                } else if (inputType === 'text' || inputType === 'number' || inputType === 'password' || inputType === 'email' || inputType === 'url' || inputType === 'tel' || inputType === 'date' || inputType === 'time' || inputType === 'color' || inputType === 'range') {
                                    mappedInnerNode.element.setAttribute('value', condition);
                                }
                            }
                        });

                        return;
                    }
                });

                // Mounting any other attribute that is not a directive.
                $FlowEngine.queueReactiveEffect({
                    effect: () => {
                        const evaluatedExpression = evaluateJSExpression(componentContext, value)

                        mappedInnerNode.element.setAttribute(translatedBindingAttribute, evaluatedExpression)
                    }
                })
            }
        })
    })

    // This is used to bind reactive variables to the DOM that are not part of a loop and evaluate expressions inside {{ }} delimiters
    const interpolationNodes = [...componentParentNode.querySelectorAll('flow-expression')].filter(element => !element.closest('Loop'))

    interpolationNodes.forEach(interpolationNode => {
        const expression = interpolationNode['data-flow-expression'] ??= interpolationNode.textContent

        const textNode = document.createTextNode('')

        interpolationNode.parentNode.insertBefore(textNode, interpolationNode)
        interpolationNode.parentNode.removeChild(interpolationNode)

        const effect = () => {
            textNode.textContent = evaluateJSExpression(componentContext, expression)
        }

        $FlowEngine.queueReactiveEffect({
            effect
        })
    })

    // Calling the onMounted lifecycle callback if it exists
    lifecycleCallbacks.onMounted?.()

    mountStyling(componentStyle, componentParentNode)

    return { componentNodeReference: componentParentNode, componentNamesUsed, nextParentContext: componentContext }
}

export function templateCompiler(componentFunction, lookUpNode, defaultEntryPointId, previousParentContext) {
    let componentNodes = []

    const componentName = componentFunction.name.toUpperCase()

    if (defaultEntryPointId) {
        componentNodes = [...lookUpNode.querySelectorAll(defaultEntryPointId)]
    } else {
        componentNodes = [...lookUpNode.querySelectorAll(`${componentName}:not([data-flow-compiled-component])`)]
    }

    componentNodes.forEach((componentNode) => {
        const { componentNodeReference, componentNamesUsed, nextParentContext } = compileComponent(componentFunction, componentNode, previousParentContext)

        if (componentNamesUsed.includes(componentName)) {
            throw new Error(`Component "${componentName}" is referenced inside its body.`)
        }

        componentNode.setAttribute('data-flow-compiled-component', true)

        componentNode.prepend(...componentNodeReference.childNodes)

        componentNamesUsed.forEach(componentName => {
            if (!$FlowEngine.componentModules[componentName]) return

            const componentModule = $FlowEngine.componentModules[componentName]

            templateCompiler(componentModule.componentFunction, componentNode, undefined, nextParentContext)
        })
    })
}
