// @flow/src/templateCompiler.js
import { evaluateJSExpression } from '../utils/evaluateJSExpression.js'
import { replaceInterpolationMarkers } from '../utils/formatTemplate.js'
import { bindForDirective } from './compiler/directives/for.js'
import { bindIfDirective } from './compiler/directives/if.js'
import { determineUsedComponents } from '../utils/determineUsedComponents.js'
import { mountStyling } from '../src/mountStyling.js'
import { getTranslatedAttributeNames, translateBindingAttribute, translateEventAttribute } from '../utils/getTranslatedAttributeNames.js'
import { reactiveVariable, createReactiveArray } from '../reactivity/reactiveVariable.js'

const BASE_REACTIVE_VARIABLE_CONFIG = {
    isStored: false,
    isRef: false,
    isComputed: false,
    isProp: false,
    influences: []
};

const REACTIVE_VARIABLE_REF_CONFIG = {
    ...BASE_REACTIVE_VARIABLE_CONFIG,
    isRef: true
};

const REACTIVE_VARIABLE_COMPUTED_CONFIG = {
    ...BASE_REACTIVE_VARIABLE_CONFIG,
    isComputed: true,
    dependsOn: []
};

const REACTIVE_VARIABLE_PROP_CONFIG = {
    ...BASE_REACTIVE_VARIABLE_CONFIG,
    isProp: true
};

const LIFECYCLE_CALLBACKS_TEMPLATE = {
    onMounted: undefined,
    onBeforeMount: undefined
}

const camelizeKebabCasedString = string => string.replace(/-./g, character => character[1].toUpperCase())

export function compileComponent(componentFunction = () => { }, componentParentNode = {}, parentContext = {}) {
    componentParentNode.$FlowComponent ??= {}

    const lifecycleCallbacks = structuredClone(LIFECYCLE_CALLBACKS_TEMPLATE)

    const properties = componentParentNode.$FlowComponent.properties = {}

    // Binding props and transforming them into reactive variables.
    let componentProperties = $FlowEngine.propertiesByComponent[componentParentNode.tagName] || [] //  - in case the parent node's tagName is 'DIV'.

    componentProperties.forEach((propertyName) => {
        const attributeExpression = componentParentNode.getAttribute(`data-flow-attribute-original-content-${propertyName}`)

        const camelCasePropertyName = camelizeKebabCasedString(propertyName)

        const propertyReactiveVariable = properties[camelCasePropertyName] = reactiveVariable(undefined, structuredClone(REACTIVE_VARIABLE_PROP_CONFIG))

        $FlowEngine.queueEffect({
            effect: () => {
                propertyReactiveVariable.value = evaluateJSExpression(parentContext, attributeExpression)
            }
        })
    })

    // Defining component's destructurable properties.
    const componentContextMethods = {
        emit: (eventName, payload, eventInitDict = { bubbles: true }) => {
            eventInitDict.detail = payload

            const customEvent = new CustomEvent(eventName, eventInitDict)

            componentParentNode.dispatchEvent(customEvent)
        },
        ref: (value) => {
            return reactiveVariable(value, structuredClone(REACTIVE_VARIABLE_REF_CONFIG))
        },
        computed: (callback) => {
            return reactiveVariable(callback, structuredClone(REACTIVE_VARIABLE_COMPUTED_CONFIG))
        },
        useStore: (storeName) => {
            const store = $FlowEngine.useStore(storeName)

            return store
        },
        watch: (source, callback, watchConfiguration) => {
            if (!source.configuration?.isReactiveVariable) return

            const variableReference = source.variableReference

            if (watchConfiguration?.immediate) {
                callback(source.value, source.value)
            }

            $FlowEngine.watchCallbacks[variableReference].push(callback)
        },
        onMounted: (callback) => {
            lifecycleCallbacks.onMounted = callback
        },
        onBeforeMount: (callback) => {
            lifecycleCallbacks.onBeforeMount = callback
        },
        props: componentParentNode.$FlowComponent.properties
    }

    // Calling a component function with the context methods. The component function returns a context object.
    // It is fine to think of it as Vue's onCreated lifecycle hook.
    const componentContext = Object.assign(properties, componentFunction(componentContextMethods))

    // Setting componentParentNode's innerHTML to the component's template and removing the 'template' property.
    const componentTemplate = componentContext.template
    delete componentContext.template

    // Invoking onBeforeMount lifecycle hook if defined.
    lifecycleCallbacks.onBeforeMount?.()

    // ! After running this instruction component gets its HTML content set.
    componentParentNode.innerHTML = replaceInterpolationMarkers(componentTemplate)

    const componentStyle = componentContext.style
    delete componentContext.style

    // Grabbing all inner elements of the componentParentNode that has just been created by innerHTML property.
    const componentNodeInnerElements = [...componentParentNode.querySelectorAll('*')]

    // Formatting every inner node in order for easier manipulation.
    let mappedInnerElements = componentNodeInnerElements.map((element) => {
        const attributes = [...element.attributes];

        return {
            element,
            attributes: getTranslatedAttributeNames(attributes)
        }
    })

    // This is so-called "loop-marking". It is used to distinguish looped elements from the rest.
    // This because loops are mounted separately from the rest of the component.
    const loopNodes = [...componentParentNode.querySelectorAll('Loop')]

    loopNodes.forEach(element => {
        const loopBlockElements = [...element.querySelectorAll('*')]

        loopBlockElements.forEach((loopBlockElement) => {
            loopBlockElement.setAttribute('data-flow-looped-element', true)
        })
    })

    // Determining component slots.
    const slotNodes = [...componentParentNode.querySelectorAll('Slot')]
    const mappedInnerSlotTags = slotNodes.map(slotNode => {
        const slotName = slotNode.getAttribute("name")

        const targetInsertTag = componentParentNode.$FlowComponent.mappedInnerInsertTags.find(insertTag => insertTag.name === slotName)

        if (targetInsertTag) {
            slotNode.prepend(...targetInsertTag.insertNode.childNodes)
        }

        return {
            element: slotNode,
            name: slotNode.getAttribute("name")
        }
    })

    componentParentNode.$FlowComponent.mappedInnerSlotTags = mappedInnerSlotTags

    // Determining which components are used in the component's template and converting their names to uppercase.
    const componentNamesUsed = determineUsedComponents(componentTemplate)

    // The famous "mounting" step. This is where the magic happens.
    mappedInnerElements.forEach((mappedInnerNode) => {
        const isComponentModule = $FlowEngine.componentModules[mappedInnerNode.element.tagName]

        if (isComponentModule) {
            mappedInnerNode.element.$FlowComponent = {}

            const insertTags = [...mappedInnerNode.element.querySelectorAll('Insert')]
            const mappedInnerInsertTags = insertTags.map(insertNode => {
                return {
                    componentNode: mappedInnerNode.element,
                    insertNode,
                    name: insertNode.getAttribute("name")
                }
            })

            mappedInnerNode.element.$FlowComponent.mappedInnerInsertTags = mappedInnerInsertTags
        }

        // Mounting event attributes that start with '@' character.
        mappedInnerNode.attributes.eventAttributes.forEach(({ name: eventAttribute, value }) => {
            mappedInnerNode.element.addEventListener(translateEventAttribute(eventAttribute), evaluateJSExpression(componentContext, value))
        })

        // Mounting binding attributes that start with ':' character.
        mappedInnerNode.attributes.bindingAttributes.forEach(({ name: bindingAttribute, value }) => {
            const translatedBindingAttribute = translateBindingAttribute(bindingAttribute)

            mappedInnerNode.element.setAttribute(`data-flow-attribute-original-content-${translatedBindingAttribute}`, value)

            // Note: for loop is mounted separately from other elements and has its own mounting (copied) step.
            if (bindingAttribute === ':for') {
                bindForDirective(mappedInnerNode, componentContext)

                return
            }

            const isNodeNotAChildOfLoopElement = !mappedInnerNode.element.getAttribute('data-flow-looped-element')

            // Mounting :if directive.
            if (isNodeNotAChildOfLoopElement) {
                if (bindingAttribute === ':if') {
                    bindIfDirective(mappedInnerNode, componentContext)

                    return
                }

                // Mounting input elements' value and checked attributes.
                const isNodeATextInput = bindingAttribute === ':value' && mappedInnerNode.element.tagName === 'INPUT' && (mappedInnerNode.element.getAttribute('type') === 'text' || !mappedInnerNode.element.hasAttribute('type'))

                if (isNodeATextInput) {
                    $FlowEngine.queueEffect({
                        effect: () => {
                            mappedInnerNode.element.value = evaluateJSExpression(componentContext, value)
                        }
                    })

                    return
                }

                const isNodeACheckboxInput = bindingAttribute === ':checked' && mappedInnerNode.element.tagName === 'INPUT' && mappedInnerNode.element.getAttribute('type') === 'checkbox';

                if (isNodeACheckboxInput) {
                    $FlowEngine.queueEffect({
                        effect: () => {
                            const condition = evaluateJSExpression(componentContext, value)

                            if (condition) {
                                mappedInnerNode.element.setAttribute('checked', condition)
                            } else {
                                mappedInnerNode.element.removeAttribute('checked')
                            }
                        }
                    })

                    return
                }

                // Mounting any other attribute that is not a directive.
                $FlowEngine.queueEffect({
                    effect: () => {
                        mappedInnerNode.element.setAttribute(translatedBindingAttribute, evaluateJSExpression(componentContext, value))
                    }
                })
            }
        })
    })

    // This is used to bind reactive variables to the DOM that are not part of a loop and evaluate expressions inside {{ }} delimiters
    const interpolationNodes = [...componentParentNode.querySelectorAll('flow-paint')].filter(element => !element.closest('Loop'))

    interpolationNodes.forEach(interpolationNode => {
        const expression = interpolationNode['data-flow-paint-expression'] ??= interpolationNode.textContent

        const textNode = document.createTextNode('')

        interpolationNode.parentNode.insertBefore(textNode, interpolationNode)
        interpolationNode.parentNode.removeChild(interpolationNode)

        const effect = () => {
            textNode.textContent = evaluateJSExpression(componentContext, expression)
        }

        $FlowEngine.queueEffect({
            effect
        })
    })

    // Calling the onMounted lifecycle callback if it exists
    lifecycleCallbacks.onMounted?.()

    mountStyling(componentStyle, componentParentNode)

    return { componentNodeReference: componentParentNode, componentNamesUsed, nextParentContext: componentContext }
}

export function templateCompiler(componentFunction, lookUpNode, defaultEntryPointId, previousParentContext) {
    let componentNodes = []

    const componentName = componentFunction.name.toUpperCase()

    if (defaultEntryPointId) {
        componentNodes = [...lookUpNode.querySelectorAll(defaultEntryPointId)]
    } else {
        componentNodes = [...lookUpNode.querySelectorAll(`${componentName}:not([data-flow-compiled-component])`)];
    }

    componentNodes.forEach((componentNode) => {
        const { componentNodeReference, componentNamesUsed, nextParentContext } = compileComponent(componentFunction, componentNode, previousParentContext)

        if (componentNamesUsed.includes(componentName)) {
            throw new Error(`Component "${componentName}" is referenced inside its body.`)
        }

        componentNode.setAttribute('data-flow-compiled-component', true)

        componentNode.prepend(...componentNodeReference.childNodes)

        componentNamesUsed.forEach(componentName => {
            if (!$FlowEngine.componentModules[componentName]) return

            const componentModule = $FlowEngine.componentModules[componentName]

            templateCompiler(componentModule.componentFunction, componentNode, undefined, nextParentContext)
        });
    })
}
