// @vind/reactivity/reactiveVariable.js
import { isPrimitive } from "../utils/isPrimitive.js";
import { REACTIVE_VARIABLE_REF_CONFIG } from "../configuration/configuration.js";

export function generateNewReference(type) {
    function getRandomNumberString(digits) {
        const min = Math.pow(10, digits - 1);
        const number = Math.trunc(Math.random() * 9 * min) + min;
        return number.toString();
    }

    const part1 = getRandomNumberString(8);
    const part2 = getRandomNumberString(8);

    return `vind-rv-${type}-id-${part1}-${part2}`;
}

const runComputedDependencesExtractor = (callback, variableReference) => {
    $VindEngine.computedDependencyExtractorRunning = true;
    $VindEngine.currentReactiveVariableReference = variableReference;

    callback();

    $VindEngine.computedDependencyExtractorRunning = false;
    $VindEngine.currentReactiveVariableReference = "";
    $VindEngine.extractedComputedDependencies.clear();
};

const getReactiveVariableType = (configuration) => {
    let type = "";

    if (configuration.isRef) {
        type = "ref";
    } else if (configuration.isComputed) {
        type = "computed";
    } else if (configuration.isProp) {
        type = "prop";
    } else if (configuration.isReactiveArray) {
        type = "reactive-array";
    } else {
        type = "unspecified";
    }

    return type;
};

const isValueAnActualObject = (value) => {
    return (
        value !== null &&
        Object.prototype.toString.call(value) === "[object Object]"
    );
};

const getReactiveVariable = (value, configuration) => {
    if (!isValueAnActualObject(value)) return value;

    const outputReactiveObject = {};

    for (const key in value) {
        const condition = isValueAnActualObject(value[key]);

        outputReactiveObject[key] = condition
            ? reactiveVariable(value[key], configuration)
            : value[key];
    }

    return outputReactiveObject;
};

const arrayMutatingMethods = [
    "push",
    "unshift",
    "pop",
    "shift",
    "splice",
    "sort",
    "reverse",
];

export const reactiveArray = (value, configuration = {}) => {
    configuration.isReactiveVariable = true;

    let type = getReactiveVariableType(configuration);

    const arrayReference = generateNewReference(type);

    $VindEngine.reactiveArraysDOMElements[arrayReference] = [];

    const reactiveArrayProxy = new Proxy(value, {
        get(target, property, receiver) {
            const isPropertyAMutativeMethod =
                typeof target[property] === "function" &&
                arrayMutatingMethods.includes(property);

            if (isPropertyAMutativeMethod) {
                return (...functionArguments) => {
                    const hasReference =
                        $VindEngine.reactiveArraysDOMElements[arrayReference];

                    if (!hasReference && (property === "push" || property === "unshift")) {
                        const reactiveElement = reactiveVariable(
                            target,
                            structuredClone(REACTIVE_VARIABLE_REF_CONFIG)
                        );

                        return Array.prototype[property].apply(
                            reactiveElement,
                            functionArguments
                        );
                    }

                    if (property === "pop") {
                        const result = target.pop();

                        $VindEngine.reactiveArraysDOMElements[arrayReference].forEach(
                            ({ HTMLCollection }) => {
                                if (HTMLCollection.length === 0) return;

                                HTMLCollection[HTMLCollection.length - 1].remove();
                            }
                        );

                        return result;
                    }

                    if (property === "shift") {
                        const result = target.shift();

                        $VindEngine.reactiveArraysDOMElements[arrayReference].forEach(
                            ({ HTMLCollection }) => {
                                if (HTMLCollection.length === 0) return;

                                HTMLCollection[0].remove();
                            }
                        );

                        return result;
                    }

                    if (property === "splice") {
                        const result = target.splice(...functionArguments);

                        $VindEngine.reactiveArraysDOMElements[arrayReference].forEach(
                            ({ HTMLCollection }) => {
                                const [start, deleteCount] = functionArguments;

                                for (let i = start; i < start + deleteCount; i++) {
                                    HTMLCollection[i].remove();
                                }
                            }
                        );

                        return result;
                    }

                    if (property === "reverse") {
                        const result = target.reverse();

                        $VindEngine.reactiveArraysDOMElements[arrayReference].forEach(
                            ({ HTMLCollection, parentLoopElement }) => {
                                if (parentLoopElement.children.length === 0) return;

                                for (let i = HTMLCollection.length - 1; i >= 0; i--) {
                                    parentLoopElement.appendChild(HTMLCollection[i]);
                                }
                            }
                        );

                        return result;
                    }

                    if (property === "sort") {
                        let result;

                        $VindEngine.reactiveArraysDOMElements[arrayReference].forEach(
                            ({ HTMLCollection, parentLoopElement }) => {
                                parentLoopElement.$VindEngine = {};
                                parentLoopElement.$VindEngine.correspondingMap = new Map();

                                Array.from(HTMLCollection).forEach((element, index) => {
                                    parentLoopElement.$VindEngine.correspondingMap.set(
                                        target[index],
                                        element
                                    );
                                });
                            }
                        );

                        $VindEngine.reactiveArraysDOMElements[arrayReference].forEach(
                            ({ parentLoopElement }) => {
                                result = Array.prototype[property].apply(
                                    target,
                                    functionArguments
                                );

                                const sortedOrderPlaceholder = document.createElement("div");

                                target.forEach((originTarget) => {
                                    sortedOrderPlaceholder.appendChild(
                                        parentLoopElement.$VindEngine.correspondingMap.get(
                                            originTarget
                                        )
                                    );
                                });

                                parentLoopElement.append(...sortedOrderPlaceholder.children);
                            }
                        );

                        return result;
                    }

                    if (property === "push") {
                        const reactiveElement = reactiveVariable(
                            functionArguments[0],
                            structuredClone(REACTIVE_VARIABLE_REF_CONFIG)
                        );

                        const result = target.push(reactiveElement);

                        $VindEngine.reactiveArraysDOMElements[arrayReference].forEach(
                            ({ parentLoopElement, mountNewLoopElement }) => {
                                const children = mountNewLoopElement(reactiveElement);

                                parentLoopElement.append(...children);
                            }
                        );

                        return result;
                    }

                    if (property === "unshift") {
                        const reactiveElement = reactiveVariable(
                            functionArguments[0],
                            structuredClone(REACTIVE_VARIABLE_REF_CONFIG)
                        );

                        const result = target.push(reactiveElement);

                        $VindEngine.reactiveArraysDOMElements[arrayReference].forEach(
                            ({ parentLoopElement, mountNewLoopElement }) => {
                                const children = mountNewLoopElement(reactiveElement);

                                parentLoopElement.prepend(...children);
                            }
                        );

                        return result;
                    }
                };
            }

            return Reflect.get(target, property, receiver);
        },
        set(target, property, newValue) {
            target[property] = newValue;

            $VindEngine.runReactiveEffects({
                reference: arrayReference,
                configuration,
            });

            return true;
        },
    });

    $VindEngine.reactiveVariables[arrayReference] = {
        variableProxy: reactiveArrayProxy,
        configuration,
    };

    return reactiveArrayProxy;
};

export const reactiveVariable = (value, configuration = {}) => {
    configuration.isReactiveVariable = true;

    let type = getReactiveVariableType(configuration);

    const variableReference = generateNewReference(type);

    $VindEngine.watchCallbacks[variableReference] = [];

    if (configuration.isComputed) {
        runComputedDependencesExtractor(value, variableReference, configuration);
    }

    configuration.isValueAnActualObject = isValueAnActualObject(value);
    configuration.isValuePrimitive = isPrimitive(value);

    const variableProxy = new Proxy(
        {
            ...(configuration.isValueAnActualObject
                ? getReactiveVariable(value, configuration)
                : { value }),
        },
        {
            get(target, property) {
                if ($VindEngine.computedDependencyExtractorRunning) {
                    $VindEngine.extractedComputedDependencies.add(variableReference);
                    configuration.influences.push(
                        $VindEngine.currentReactiveVariableReference
                    );

                    return;
                }

                if ($VindEngine.dependencyExtractorRunning) {
                    $VindEngine.extractedDependencies.add(variableReference);
                }

                const propertyValue = target[property];

                return typeof propertyValue === "function"
                    ? propertyValue()
                    : propertyValue;
            },

            set(target, property, newValue) {
                const previousValue = target[property];

                if (configuration.isValuePrimitive && newValue === previousValue) {
                    return true;
                }

                target[property] = newValue;

                const watchCallbacks = $VindEngine.watchCallbacks[variableReference];

                if (watchCallbacks.length > 0) {
                    watchCallbacks.forEach((watchSignal) =>
                        watchSignal(newValue, previousValue)
                    );
                }

                const influencedReferences = configuration.influences;

                influencedReferences.forEach((influencedReference) => {
                    const computedWatchCallbacks =
                        $VindEngine.watchCallbacks[influencedReference];

                    const computedReactiveReference =
                        $VindEngine.reactiveVariables[influencedReference].variableProxy;

                    computedWatchCallbacks.forEach((watchSignal) =>
                        watchSignal(
                            computedReactiveReference.value,
                            computedReactiveReference.value
                        )
                    );
                });

                $VindEngine.runReactiveEffects({
                    reference: variableReference,
                    configuration,
                });

                return true;
            },
        }
    );

    $VindEngine.reactiveVariables[variableReference] = {
        variableProxy,
        configuration,
    };

    return variableProxy;
};
