// @vind/index.js
import { templateCompiler } from "./src/compiler/templateCompiler.js";
import { showErrorOverlay } from "./utils/showErrorOverlay.js";
// import { showLoadingOverlay } from "./utils/showLoadingOverlay.js"
import { generateStore } from "../@vind/store/generateStore.js";

const DEFAULT_ENTRY_POINT_COMPONENT_NAME = "App";
const DEFAULT_ENTRY_POINT_ELEMENT_ID = "#app";

const callWithErrorOverlay = (callback) => {
    try {
        callback();
    } catch (error) {
        showErrorOverlay(error);
        throw new Error(error);
    }
};

export const Vind = {
    render(callback) {
        callback()

        $VindEngine.storeGenerationQueue.forEach(callWithErrorOverlay);

        callWithErrorOverlay(() => {
            const upperCaseEntryPointComponentName = DEFAULT_ENTRY_POINT_COMPONENT_NAME.toUpperCase();

            const { componentFunction: defaultEntryPointComponentFunction } = $VindEngine.componentModules[upperCaseEntryPointComponentName];

            templateCompiler(
                defaultEntryPointComponentFunction,
                document.body,
                undefined,
                DEFAULT_ENTRY_POINT_COMPONENT_NAME,
                { defaultEntryPointId: DEFAULT_ENTRY_POINT_ELEMENT_ID, target: 'general' }
            );
        });
    },
    defineComponents(componentModules = []) {
        componentModules.forEach((componentModule) => {
            $VindEngine.defineComponentModule(componentModule);
        });
    },
    defineStores(stores = []) {
        stores.forEach((storeFunction) => {
            $VindEngine.defineStore(storeFunction);
        });
    },
};

window.$VindEngine = (() => {
    const reactiveEffects = [];
    const lastInfluencedEffectReference = "";

    const reactiveVariables = {};
    const watchCallbacks = {};
    const stores = {};
    const storeGenerationQueue = [];

    const propertiesByComponent = {};
    const eventsByComponent = {};
    const templateByComponent = {};
    const styleByComponent = {};
    const componentsWithStyleMounted = new Set()

    const componentModules = {};

    const computedDependencyExtractorRunning = false;
    const extractedComputedDependencies = new Set();
    const currentReactiveVariableReference = "";

    const dependencyExtractorRunning = false;
    const extractedDependencies = new Set();

    const reactiveArraysDOMElements = {};

    // Effects
    const queueReactiveEffect = (
        newReactiveEffect,
        configuration = { runOnQueue: true, predefinedReferences: [] }
    ) => {
        $VindEngine.dependencyExtractorRunning = true;

        newReactiveEffect.references = new Set();

        const predefinedReferences = configuration.predefinedReferences

        if (configuration.runOnQueue === undefined || configuration.runOnQueue) {
            newReactiveEffect.effect();
        }

        if ($VindEngine.extractedDependencies.size > 0) {
            newReactiveEffect.references = new Set([...$VindEngine.extractedDependencies]);

            $VindEngine.extractedDependencies.clear()
        }

        if (predefinedReferences?.length > 0) {
            predefinedReferences.forEach((reference) => {
                newReactiveEffect.references.add(reference);
            });
        }

        newReactiveEffect.references = Array.from(newReactiveEffect.references);

        reactiveEffects.push(newReactiveEffect);

        $VindEngine.dependencyExtractorRunning = false;
    };

    const runReactiveEffects = ({
        reference: reactiveVariableReference,
        configuration = {},
    }) => {
        if ($VindEngine.lastInfluencedEffectReference === reactiveVariableReference) return;

        $VindEngine.lastInfluencedEffectReference = reactiveVariableReference;

        reactiveEffects.forEach(({ references, effect }) => {
            references.forEach((effectReference) => {
                if (effectReference === reactiveVariableReference) {
                    effect();

                    configuration.influences.forEach((influencedReference) => {
                        const influencedReferenceConfiguration = reactiveVariables[reactiveVariableReference].configuration || {};

                        runReactiveEffects({
                            reference: influencedReference,
                            configuration: influencedReferenceConfiguration,
                        });
                    });
                }
            });
        }
        );

        $VindEngine.lastInfluencedEffectReference = "";
    };

    const defineComponentModule = (componentModule, customComponentName = "") => {
        // TODO: This 'find' instruction is not checking whether the found 'function' is an actual component.
        let componentFunction = Object.values(componentModule).find(
            (value) => typeof value === "function"
        );

        const upperCaseComponentName = (customComponentName || componentFunction.name).toUpperCase();

        $VindEngine.templateByComponent[upperCaseComponentName] =
            componentModule["Template"] || ``;

        $VindEngine.styleByComponent[upperCaseComponentName] =
            componentModule["Style"] || ``;

        $VindEngine.propertiesByComponent[upperCaseComponentName] =
            componentModule["Props"] || [];

        $VindEngine.eventsByComponent[upperCaseComponentName] =
            componentModule["Emits"] || [];

        componentModules[upperCaseComponentName] = {
            componentFunction,
            componentName: customComponentName || componentFunction.name,
        };
    };

    const defineStore = (storeFunction) => {
        const storeName = storeFunction.name;

        const storeGenerationFunction = generateStore(storeName, storeFunction);

        storeGenerationQueue.push(storeGenerationFunction);
    };

    return {
        componentModules,
        defineComponentModule,

        templateByComponent,
        propertiesByComponent,
        eventsByComponent,
        styleByComponent,
        componentsWithStyleMounted,

        reactiveVariables,
        reactiveArraysDOMElements,
        stores,
        storeGenerationQueue,
        defineStore,

        effects: reactiveEffects,
        lastInfluencedEffectReference,
        watchCallbacks,
        queueReactiveEffect,
        runReactiveEffects,

        computedDependencyExtractorRunning,
        extractedComputedDependencies,
        currentReactiveVariableReference,

        dependencyExtractorRunning,
        extractedDependencies,
    };
})();

// * Temporary implemetation of the router aka. "the real router"
const routerStore = ({ ref }) => {
    const path = ref("/");

    return {
        path,
    };
};

const routerStoreGenerationFunction = generateStore("routerStore", routerStore);
$VindEngine.storeGenerationQueue.push(routerStoreGenerationFunction);
